### ISSUE TITLE
File parsing has seemingly changed in V4 (BC break)

### ISSUE THREAD

**Issue by jameshalsall** [serverless/serverless#12910]
### Issue description

# Bug

We have a monorepo. There is a common `includes.yml` file that we import into many different `serverless.yml` files. 

```yaml
# includes.yml
stage: ${opt:stage, 'staging'}
```

```yaml
# application serverless.yml

custom: ${file(../../../tools/serverless/includes.yml)}

provider:
  stage: ${self:custom.stage}
```

When I run this in V4, I see
```
✖ Error: Cannot resolve '${self:custom.stage}' variable at 'provider.stage'. No value is available for this variable, and no default value was provided. Please check your variable definitions or provide a default value.
    at _ResolverManager.resolve (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:57377)
    at async #handlePlaceholderNode (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:55807)
    at async file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:53799
    at async processNodeAndHandleCompletion (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:43828)
```

However, in V3 this works OK. 

If we update the application's `serverless.yml` file to the below...
```yaml
# application serverless.yml

custom: ${file(../../../tools/serverless/includes.yml)}

provider:
  stage: ${self:custom.stage, 'staging'}
```

Then everything is OK. 

If we then change the `custom` include file path to something that is invalid, e.g.
```yaml
# application serverless.yml

custom: ${file(../../../foo/includes.yml)}

provider:
  stage: ${self:custom.stage, 'staging'}
```

Then it complains validly that the `custom` variable cannot be resolved:
```
✖ Error: Cannot resolve '${file(../../../foo/includes.yml)}' variable at 'custom'. No value is available for this variable, and no default value was provided. Please check your variable definitions or provide a default value.
    at _ResolverManager.resolve (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:57377)
    at async #handlePlaceholderNode (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:55807)
    at async file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:53799
    at async processNodeAndHandleCompletion (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:43828)
```

If I then remove the default value on the `provider.stage` we arrive back to the original error of
```
✖ Error: Cannot resolve '${self:custom.stage}' variable at 'provider.stage'. No value is available for this variable, and no default value was provided. Please check your variable definitions or provide a default value.
    at _ResolverManager.resolve (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:57377)
    at async #handlePlaceholderNode (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:55807)
    at async file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:53799
    at async processNodeAndHandleCompletion (file:///Users/jameshalsall/.serverless/releases/4.4.10/package/dist/sf-core.js:693:43828)
```

# Summary
To me, this proves that in V4 the `serverless.yml` file is being parsed before includes are being resolved. Maybe the includes are being resolved in a second pass or something? It's a fundamental issue though and a BC break, unless I'm missing something.

### Context

# Service Overview
- Serverless Framework Version: 4.4.10
- Command: support --summary true

(this command does not work properly it seems 😀)

**Comment by joe-price-jt** [comment]
Is there any update on this? I am running into the same issue with a similar setup.

In my case, I have a common `parameters.yml` file that contains custom parameters and variables for use across multiple services. This file defines a `params` block and `custom` block, and are used to populate the `params` and `custom` fields in the main `serverless.yml` file:

```yaml
# parameters.yml
params:
  dev:
    region: 'eu-west-2'

custom:
  region: ${param:region}
```

```yaml
# serverless.yml of a particular service
service: some-service

frameworkVersion: '4'

custom: ${file(../parameters.yml):custom}
params: ${file(../parameters.yml):params}
  ...
```

I have properties defined in the `params` block of `parameters.yml` that reference properties defined in the `custom` block of the same file.

This setup worked in V3, but since moving to V4, I have encountered a similar error to the listed in the issue:

```
 Error: Cannot resolve '${param:region}' variable at 'custom.region'. No value is available for this variable, and no default value was provided. Please check your variable definitions or provide a default value.
    at _ResolverManager.resolve (file:///Users/joeprice/.serverless/releases/4.4.14/package/dist/sf-core.js:693:57114)
    at async #handlePlaceholderNode (file:///Users/joeprice/.serverless/releases/4.4.14/package/dist/sf-core.js:693:55518)
    at async file:///Users/joeprice/.serverless/releases/4.4.14/package/dist/sf-core.js:693:53510
    at async processNodeAndHandleCompletion (file:///Users/joeprice/.serverless/releases/4.4.14/package/dist/sf-core.js:693:44152)
```

It seems that the file parsing runs into an issue when defining the `custom` block of a `serverless.yml` file from a file import specifically. I run into no errors when defining the `custom` block directly in the `serverless.yml` file.

Performed using Serverless Framework version 4.4.15.

**Comment by czubocha** [comment]
@joe-price-jt Can you confirm if this issue persists in the latest version (4.4.16)? Based on your stack trace, you’re using 4.4.14, and I was able to reproduce the error in that version. However, it seems to be resolved in 4.4.15. Here’s the config I used to reproduce:

```
#parameters.yml
params:
  dev:
    region: 'eu-west-2'

custom:
  region: ${param:region}
```

```
service: some-service

custom: ${file(parameters.yml):custom}
params: ${file(parameters.yml):params}


provider:
  name: aws
  region: ${self:custom.region}

functions:
  hello:
    handler: handler.hello
    description: ${self:custom.region}
    runtime: nodejs22.x
```

**Comment by joe-price-jt** [comment]
Hey @czubocha! I can confirm that the issue no longer persists in the latest version 🥳 Thanks for getting back to me on this!

**Comment by czubocha** [comment]
Hey @jameshalsall, I wanted to let you know that your use case should now be addressed in [v4.4.17](https://github.com/serverless/serverless/releases/tag/v4.4.17). I’m closing this issue for now, but if you encounter any further problems or have additional feedback, feel free to reopen or let us know!
